File: examples/python/basic.py
#!/usr/bin/env python3
"""Basic StacyVM Python SDK usage.

Demonstrates:
  - Connecting to a StacyVM server
  - Spawning a sandbox
  - Executing a command
  - Writing and reading a file
  - Destroying the sandbox

Prerequisites:
  pip install stacyvm
  # StacyVM server running at localhost:7423
"""

from stacyvm import Client


def main() -> None:
    # Connect to a local StacyVM server (default port 7423).
    # Pass api_key="..." if authentication is enabled.
    with Client("http://localhost:7423") as client:

        # Check the server is healthy before doing real work.
        health = client.health()
        print(f"Server status: {health['status']}  version: {health.get('version', 'n/a')}")

        # Spawn a sandbox using the default provider and image.
        sandbox = client.spawn(image="alpine:latest", ttl="10m")
        print(f"Spawned sandbox: {sandbox.id}  state={sandbox.state}")

        # Get a preview URL (e.g. for a web server running on port 3000).
        preview_url = sandbox.get_preview_url(3000)
        print(f"Live Preview URL: {preview_url}")

        # Execute a command inside the sandbox.
        result = sandbox.exec("echo 'Hello from StacyVM!'")
        print(f"Exit code: {result.exit_code}")
        print(f"Stdout: {result.stdout.strip()}")

        # Write a file into the sandbox.
        sandbox.write_file("/tmp/greeting.txt", "Hello, world!\n")
        print("Wrote /tmp/greeting.txt")

        # Read it back.
        content = sandbox.read_file("/tmp/greeting.txt")
        print(f"Read back: {content.strip()!r}")

        # List files in /tmp.
        files = sandbox.list_files("/tmp")
        print(f"Files in /tmp: {[f['path'] for f in files]}")

        # Clean up.
        sandbox.destroy()
        print(f"Sandbox {sandbox.id} destroyed.")


if __name__ == "__main__":
    main()

File: examples/js/basic.ts
/**
 * Basic StacyVM TypeScript SDK usage.
 *
 * Demonstrates:
 *   - Connecting to a StacyVM server
 *   - Spawning a sandbox
 *   - Executing a command
 *   - Writing and reading a file
 *   - Using withSandbox for automatic cleanup
 *
 * Prerequisites:
 *   npm install stacyvm    (or link the local SDK)
 *   # StacyVM server running at localhost:7423
 *
 * Run with:
 *   npx tsx examples/js/basic.ts
 */

import { Client } from "stacyvm";

async function main(): Promise<void> {
  // Connect to a local StacyVM server on the default port.
  // Pass apiKey: "..." if authentication is enabled.
  const client = new Client({ host: "localhost", port: 7423 });

  // Check server health.
  const health = await client.health();
  console.log(`Server status: ${health.status}  version: ${health.version}`);

  // List available providers.
  const providers = await client.providers();
  for (const p of providers) {
    console.log(`  provider: ${p.name}  healthy=${p.healthy}  default=${p.default}`);
  }

  // --- Manual lifecycle --------------------------------------------------

  console.log("\n=== Manual lifecycle ===");
  const sandbox = await client.spawn({
    image: "alpine:latest",
    ttl: "10m",
  });
  console.log(`Spawned: ${sandbox.id}  state=${sandbox.state}`);

  // Get a preview URL (e.g. for a web server running on port 3000).
  const previewUrl = sandbox.getPreviewUrl(3000);
  console.log(`Live Preview URL: ${previewUrl}`);

  // Execute a command.
  const result = await sandbox.exec("echo 'Hello from StacyVM!'");
  console.log(`Exit code: ${result.exit_code}`);
  console.log(`Stdout: ${result.stdout.trim()}`);

  // Write a file.
  await sandbox.writeFile("/tmp/greeting.txt", "Hello, world!\n");
  console.log("Wrote /tmp/greeting.txt");

  // Read it back.
  const content = await sandbox.readFile("/tmp/greeting.txt");
  console.log(`Read back: ${JSON.stringify(content.trim())}`);

  // List files.
  const files = await sandbox.listFiles("/tmp");
  console.log(`Files in /tmp: ${files.map((f) => f.name).join(", ")}`);

  // Destroy.
  await sandbox.destroy();
  console.log(`Sandbox ${sandbox.id} destroyed.\n`);

  // --- withSandbox (automatic cleanup) -----------------------------------

  console.log("=== withSandbox ===");
  const output = await client.withSandbox(
    { image: "alpine:latest", ttl: "5m" },
    async (sb) => {
      await sb.writeFile("/app/main.sh", '#!/bin/sh\necho "Computed: $((6 * 7))"');
      const res = await sb.exec("sh /app/main.sh");
      return res.stdout.trim();
    },
  );
  console.log(`Result: ${output}`);
  console.log("Sandbox automatically destroyed.");
}

main().catch((err) => {
  console.error("Fatal:", err);
  process.exit(1);
});
